src/components/App/ImportStudyList.js

```diff
@@ -0,0 +1,122 @@
+import React, { PureComponent } from 'react';
+import {
+    Button,
+    Dialog,
+    DialogActions,
+    DialogContent,
+    DialogContentText,
+    DialogTitle,
+    TextField,
+} from '@material-ui/core';
+import { queryWebsoc } from '../../helpers';
+import store from '../../stores/RightPaneStore';
+import { addCourse, clearSchedules, openSnackbar } from '../../actions/AppStoreActions';
+import AppStore from '../../stores/AppStore';
+import { GetApp } from '@material-ui/icons';
+
+class ImportStudyList extends PureComponent {
+    state = {
+        isOpen: false,
+        selectedTerm: null,
+        studyListText: '',
+    };
+
+    handleError = (index) => {
+        clearSchedules([index]);
+        openSnackbar('error', 'An error occurred while trying to import the Study List.');
+    };
+
+    handleOpen = () => {
+        this.setState({ isOpen: true, selectedTerm: store.getFormData().term });
+    };
+
+    handleClose = (doImport) => {
+        this.setState({ isOpen: false }, () => {
+            document.removeEventListener('keydown', this.enterEvent, false);
+            this.setState({ studyListText: '' });
+        });
+        if (doImport) {
+            document.removeEventListener('keydown', this.enterEvent, false);
+            if (!(this.state.studyListText && this.state.studyListText.match(/\d{5}/g))) {
+                openSnackbar('error', 'Cannot import an empty/invalid Study List.');
+                return;
+            }
+            const currIndex = AppStore.getCurrentScheduleIndex();
+            try {
+                Promise.all(
+                    this.state.studyListText
+                        .split('\n')
+                        .map((line) => line.match(/\d{5}/g))
+                        .filter((id) => id)
+                        .flat()
+                        .map((sectionCode) => queryWebsoc({ term: this.state.selectedTerm, sectionCodes: sectionCode }))
```

Comment:
Instead of querying websoc/PP API for each section code, since a studylist will have all courses in the same term, you can build a list of section codes and submit them in a single request. Keep in mind though, this will limit you to ~8 section codes (last time that I checked), so chunk your codes into groups of 8 and submit them in more requests if you need to.